### ISSUE TITLE
Promises from children added by `run` are valid but not outside

### ISSUE THREAD

**Issue by paboldin** [DataBiosphere/toil#1934]
Here is the sample codes, first one fails and second works. May you guys please explain why?

Fails:
```python
from toil.common import Toil
from toil.job import Job

class Root(Job):
    pass

class B(Job):
    pass

if __name__ == '__main__':
    parser = Job.Runner.getDefaultArgumentParser()
    options = parser.parse_args()

    rootJob = Root()

    b = B()
    rootJob.addChild(b)
    rootJob.promised = b.rv()

    with Toil(options) as toil:
        toil.start(rootJob)
```

Works:
```python
from toil.common import Toil
from toil.job import Job

class Root(Job):
    def run(self, fileStore):
        b = B()

        self.promised = b.rv()
        self.addChild(b)

class B(Job):
    pass

if __name__ == '__main__':
    parser = Job.Runner.getDefaultArgumentParser()
    options = parser.parse_args()

    rootJob = Root()

    with Toil(options) as toil:
        toil.start(rootJob)
```

┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/TOIL-212)
┆Issue Number: TOIL-212

**Comment by ejacox** [comment]
Neither case makes sense. The root job runs first, then the B job. If the root job needs the results of the B job, then the root should be the child.

**Comment by paboldin** [comment]
Thanks for the answer.

My use case is following: I spawn childs from a job and then collect childs' return values to an object that parent returns. Is there any correct way to collect childs' `rv`s or is my scheme broken completely?

**Comment by ejacox** [comment]
Your root job should be split - one part runs before the children, the other after. Take a look at our sort example: http://toil.readthedocs.io/en/3.11.0/gettingStarted/quickStart.html#a-more-real-world-example.

The setup() function runs first and creates child jobs (down()). The up() function collects the results of the children.

In general, the workflow can not contain cycles (results delivered to ancestors).

**Comment by paboldin** [comment]
Here is the code, may you please take a glance?

https://github.com/paboldin/libcare/blob/0dc61529b5317468642b6b4474a225f56c261f31/scripts/toil/pkgbuild.py#L68

Basically, I'm trying to avoid chaining jobs from inside other jobs as this makes the whole thing a little bit implicit. In my case I'm returning a `self._storage` by overriding `_run` method. It passes storage update by childs down to the followOns.

How wrong is that? I don't see much of class-based jobs in bd2k code, and the code is mostly of tail-call-chained functions, so I would really appreciate guidance here.

**Comment by ejacox** [comment]
The code is hard to follow without comments or descriptions.  A diagram describing the job relations would be helpful. From a glance, I find that WorkflowToFileStoreProxy is strange or passing the workflow as an argument. Redoing those parts might help.

**Comment by adamnovak** [comment]
Child jobs actually run *after* the code of the job that launches them. You can get promises for their return values with `.rv()`, but you can't really do any manipulation of them (like concatenating them). You also can't run code on them, like you can with JavaScript's `then()`.

You can put them in a list and return the list, though; Toil will fix up the list and replace the promises with the actual values.

Hopefully that helps. Please re-open this issue if you have more questions about the promise system.
